spezi-invitation-code/index.js

```diff
@@ -0,0 +1,101 @@
+//
+// This source file is part of the Stanford Spezi open-source project
+//
+// SPDX-FileCopyrightText: 2022 Stanford University and the project authors (see CONTRIBUTORS.md)
+//
+// SPDX-License-Identifier: MIT
+//
+
+const admin = require("firebase-admin");
+const {https} = require("firebase-functions/v2");
+const {FieldValue} = require("firebase-admin/firestore");
+
+/**
+ * Invitation code verifier for Spezi Firebase.
+ */
+class InvitationCodeVerifier {
+  /**
+   * Create an Invitation Code Verifier.
+   * @param {string} [invitationCodePath="invitationCodes"] - The path in Firestore where invitation codes are stored.
+   * @param {string} [userPath="users"] - The path in Firestore where user data is stored.
+   * @param {RegExp} [invitationCodeRegex=null] - The regex to validate invitation codes. If null, no validation is performed.
+   */
+  constructor(invitationCodePath = "invitationCodes", userPath = "users", invitationCodeRegex = null) {
+    this.firestore = admin.firestore();
+    this.invitationCodePath = invitationCodePath;
+    this.userPath = userPath;
+    this.invitationCodeRegex = invitationCodeRegex;
+  }
+
+  /**
+   * Enroll a user in a study using an invitation code.
+   * @param {string} userId - The ID of the user to enroll.
+   * @param {string} invitationCode - The invitation code to use for enrollment.
+   * @throws Will throw an error if the userId or invitationCode is invalid, not found or already used, or if the user is already enrolled.
+   */
+  async enrollUserInStudy(userId, invitationCode) {
+    if (!userId) {
+      throw new https.HttpsError(
+          "invalid-argument",
+          "The function must be called with a valid 'userId' input.",
+      );
+    }
```

Comment:
I would suggest that we get the uid from the request auth section and not from an input by the user. Would make this even more secure and we can rely on Firebase to handle the auth for us.

Just to be sure that users do this properly we should probably add an overload/alternative that takes in a request auth object and the invitationCode? Would be great if we can guide users to that.

```js
if (!request.auth.uid) {
  throw new HttpsError("...", "...");
}

const userId = request.auth.uid;
```

Reply:
The auth context contains the relevant authentication information parsed based on the user that called the function: https://firebase.google.com/docs/reference/functions/2nd-gen/node/firebase-functions.https.callablerequest.md#httpscallablerequestauth

It contains the token and uid: https://firebase.google.com/docs/reference/functions/2nd-gen/node/firebase-functions.tasks.authdata

Therefore we could ask a user to pass in the `request` instead of the `userId`. 

In combination with a TS interface we might make it more clearer what types we are expecting.

Reply:
Nice! Thanks for the additional information. Guess I am still trying to figure out what I have to do to spin up such a `request` object in `invitationCodeVerifier.test.js` and pass that in. The line between client and server code seems to blend in my mind when writing the unit tests and figuring out which modules to import. Maybe I have to simulate calling a cloud function in the unit tests?

Reply:
I think generating a JS Request object that just contains the auth id should be sufficient. Could be a Firebase Request type but as we don't yet fully enforce types with TS here it could even be an object that is structured in the same way?